Trace overview queries

jasonrhodes opened the pull request:
Closes elastic/kibana#23438

<!--
Thank you for your interest in and contributing to Kibana! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md)?
- If submitting code, have you included unit tests that cover the changes?
- If submitting code, have you tested and built your code locally prior to submission with `yarn test && yarn build`?
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
-->

jasonrhodes commented:
Closes #23438

elasticmachine commented:
Pinging @elastic/apm-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3618/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3620/)

jasonrhodes commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3622/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3624/)

sqren commented on x-pack/plugins/apm/public/components/shared/ImpactBar/index.tsx:
```diff
@@ -8,13 +8,25 @@ import { EuiProgress } from '@elastic/eui';
 
-export function ImpactBar({ value, ...rest }: Props) {
+export function ImpactBar({ value, range = [], ...rest }: Props) {
+  const maxImpact = Math.max(0, ...range.map(({ impact }) => impact));
```
I think `max={max}` makes a little more sense to me than `range={items}`. When looking at the component from the outside, I was not sure what `range` did. 

There is also the perf consideration that the max will be calculated for every single item.

sqren replied:
> The same calculation has to happen in one place or the other, though, whether it gets defined outside this component each time it's used or just defined the one time here, which is why I opted to move it in

I don't think that's right. Right now `maxImpact` is defined inside `ImpactBar`, which will be rendered for every item in the list. So if there are 100 items, it will be calculated 100 times. If `maxImpact` is defined outside `ImpactBar` it will only be calculated once, regardless how many items there are.

Additionally I think we need a `minImpact` also. We used to calculate it like this (slightly convoluted)
https://github.com/sqren/kibana/blob/6044cfb463473848af3225d97f95754be5538b5a/x-pack/plugins/apm/public/store/reactReduxRequest/transactionList.js#L17-L21

The `minImpact` exaggerates small differences for high values. 
Eg. If all values are between 90 and 100
Sample value: 91
Previous impact: 10%
New impact: 91%

Sample value: 95
Previous impact: 50%
New impact: 95%

Perhaps it would be simpler to calculate this upfront in node?
https://github.com/elastic/kibana/blob/master/x-pack/plugins/apm/server/lib/transactions/get_top_transactions.js#L87

sqren replied:
Something like this would work:
```js
function getRelativeImpact(value, min, max) {
  return (value - min) / (max - min)
}
```

jasonrhodes replied:
I see what you're saying now about having to calculate the same value for every item, good catch. The need for a minimum makes sense too. I'll work on something.

sqren commented on x-pack/plugins/apm/public/components/app/TraceOverview/TraceList.tsx:
```diff
@@ -75,19 +73,19 @@ const TRACE_COLUMNS = [
   return (
     <ManagedTable
-      columns={TRACE_COLUMNS}
+      columns={getTraceColumns({ items })}
```
Probably not a big issue, but before `TRACE_COLUMNS ` was defined outside the render function, and thus `ManagedTable` would not re-render whenever `TraceList` re-rendered. Now `getTraceColumns` returns a new object every time `TraceList` re-renders , which will cause `ManagedTable` to re-render also.

sqren commented on x-pack/plugins/apm/public/services/rest/apm.js:
```diff
@@ -64,6 +64,17 @@ export async function loadServiceDetails({ serviceName, start, end, kuery }) {
+    }
+  });
+}
+
```
Microscopic nit:
It just occurred to me that it might make sense to group the methods together by endpoints, eg. `loadTraceList` and `loadTrace`. Either in a separate file like `rest/apm/traces.ts` (to mirror the server), or just closer together in this file.

jasonrhodes replied:
Waiting on re-ordering these for now, but I agree with you. Can clean these up later when we're not both in the files at the same time.

sqren replied:
totally fine. Not necessary in this PR.

sqren commented on x-pack/plugins/apm/server/lib/traces/get_top_traces.js:
```diff
@@ -0,0 +1,104 @@
+        bool: {
+          must: {
+            exists: {
+              field: 'trace.id' // todo: move to constant? or TS?
```
I thought we could get rid of the constants.js but for these cases I think the constants still make sense. 
I was actually going to make a test that checks the constants against our typescript interfaces, so we are sure they don't get out of sync.

so tldr: unless you know a way to do this with typescript, use a constant.

sqren commented on x-pack/plugins/apm/server/lib/traces/get_top_traces.js:
```diff
@@ -0,0 +1,104 @@
+  });
+
+  // Sort results by impact - needs to be desc, hence the reverse()
+  return sortBy(results, o => o.impact).reverse();
```
We do actually have _.orderBy that supports sort order:

https://github.com/elastic/kibana/blob/4126d51044d8e85f80548e21e1615d20ded715d8/x-pack/package.json#L148

jasonrhodes replied:
Ah yeah this was just all moved over from the get_top_transactions file

https://github.com/elastic/kibana/blob/2c4e263982bd0517a1de4eda970898f76c9e5aed/x-pack/plugins/apm/server/lib/transactions/get_top_transactions.js#L100

I was considering making "get top traces" (maybe renamed "get trace list") to just use get transactions under the hood, with maybe an option to adjust the query slightly, since so much of those files are now duplicate. Thoughts?

sqren replied:
Ah, I love pointing out my own mistakes in other people's PRs :D

sqren commented:
@jasonrhodes regarding
https://github.com/elastic/kibana/pull/23605#issuecomment-425402661

I think you need the "closes" keyword in the PR description for it to work.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3732/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3828/)

jasonrhodes commented:
Ping @sqren -- I think this addresses most of the feedback. Went with a `<RelativeImpactGroup>` component that provides the overall values for the group and a curried `<RelativeImpactBar>` component that already has the correct `max` value applied for you.

Also made a `getTransactionList` function that both `getTopTransactions` and `getTopTraces` now use. Decided those don't need to be sorted by impact at all since the tables that use them are also applying an "initial sort" by impact, desc.

sqren commented on x-pack/plugins/apm/public/components/shared/ImpactBar/index.tsx:
```diff
@@ -8,13 +8,61 @@ import { EuiProgress } from '@elastic/eui';
+      relativeMax
+    }
+  });
 }
```
I think this component seem a overly complicated. Why not just calculated it in Node, and just use `EuiProgress` directly? Then we can also avoid the render prop. I prefer render props over HOC but but they both add complexity, and I don't think it's necessary in this case.

jasonrhodes replied:
If we calculate it in node then we have to pass along the "max" value somehow. I didn't see a simple and clean way of doing that without having to calcuate it multiple times.

sqren replied:
Do we need the max value? Can't we just calculate the relative impact as percentage, and provide that to `EuiProgress`? Implicitly the max will be 100 but we don't have to pass that along.

sqren commented on x-pack/plugins/apm/public/components/app/TraceOverview/TraceList.tsx:
```diff
@@ -35,7 +37,12 @@ interface TraceItem {
+  noItemsMessage: any;
+}
+
+const traceListColumns: ITableColumn[] = [
```
The "I" prefix still seems pretty old-school to me, but I think it's the most pragmatic solution to add it, to avoid conflicts between component names and interfaces. Should we make it a rule to have it for all interfaces?

There is a ts-lint rule: https://palantir.github.io/tslint/rules/interface-name/

jasonrhodes replied:
I'm in favor of this, but perhaps during a general "TypeScript cleanup" phase post DT?

sqren replied:
Sure!

sqren commented on x-pack/plugins/apm/public/components/app/TraceOverview/view.tsx:
```diff
@@ -23,7 +24,7 @@ export function TraceOverview(props: Props) {
       <TraceListRequest
         urlParams={urlParams}
-        render={({ data }: { data: object[] }) => (
+        render={({ data }: { data: ImpactItem[] }) => (
```
Doesn't `TraceListRequest` return a list of trace items, and not impact items?

sqren replied:
Btw. it would be cool if we could add an interface to the node endpoints, and re-use that here! :)

sqren commented on x-pack/plugins/apm/public/components/app/TraceOverview/view.tsx:
```diff
@@ -23,7 +24,7 @@ export function TraceOverview(props: Props) {
       <TraceListRequest
         urlParams={urlParams}
-        render={({ data }: { data: object[] }) => (
+        render={({ data }: { data: ImpactItem[] }) => (
```
Doesn't `TraceListRequest` return a list of trace items, and not impact items?

sqren commented on x-pack/plugins/apm/server/lib/traces/get_top_traces.js:
```diff
@@ -0,0 +1,50 @@
+      tracesPerMinute: tpm,
+      impact
+    })
+  });
```
I think this level of abstraction makes the code harder to read. I'd prefer less DRY code which is more readable - especially when the duplication is only in two places.

Perhaps you can extract just the aggregation?

sqren replied:
Btw. I can totally related to this. When I first started in Elastic I wrote a DSL to get rid of the duplication in ES queries. I definitely went way further than you did :)
https://github.com/elastic/x-pack-kibana/pull/2653

jasonrhodes replied:
Yeah, I find this easier to understand and follow because it makes it clear what's different and what's not and helps confirm that traces are transactions, but I know abstraction has a trade-off cost. I don't mind repeating things, but I'm worried we're introducing a fair amount of code to retrieve "traces" that is actually just getting transactions with slightly different criteria to query for. What I think I'd really prefer is removing the `transformResult` altogether and aligning the results from both queries to have the same keys, which would make this even more streamlined.

sqren replied:
I see. I think if you can get rid of `transformResult` it would make it a lot more readable, and I'd be good with it 👍

jasonrhodes commented on x-pack/plugins/apm/common/constants.js:
```diff
@@ -17,6 +17,7 @@ export const TRANSACTION_RESULT = 'transaction.result';
 export const TRANSACTION_NAME = 'transaction.name';
 export const TRANSACTION_ID = 'transaction.id';
 export const TRANSACTION_SAMPLED = 'transaction.sampled';
+export const PARENT_ID = 'parent.id'; // deprecated
```
Yep. Was also looking at it being in the "must not exist" section of the query but I remember why now.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3984/)

sqren commented on x-pack/plugins/apm/typings/TransactionGroup.ts:
```diff
@@ -0,0 +1,17 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export interface ITransactionGroup {
```
I can't figure out if this should live here, or in the `transaction_query` file.

jasonrhodes replied:
This is an interface that's shared between server and client so it seems to make sense to live here.

sqren replied:
Good point 👍

sqren commented on x-pack/plugins/apm/server/lib/helpers/transaction_query.ts:
```diff
@@ -0,0 +1,119 @@
+/*
```
`transaction_query.ts` is perhaps a bit too generic, since we also have [`get_transaction`](https://github.com/elastic/kibana/blob/master/x-pack/plugins/apm/server/lib/transactions/get_transaction.js) that doesn't use this.

jasonrhodes replied:
I'll call it transaction_group_query

sqren commented on x-pack/plugins/apm/server/lib/helpers/transaction_query.ts:
```diff
@@ -0,0 +1,119 @@
+}: {
+  response: ITransactionGroupQueryResult;
+  durationMinutes: number;
+}) {
```
I think `prepareTransactionGroups` should just take `buckets` and not the full response. Then you can get rid of `ITransactionGroupQueryResult`.

btw. I added elasticsearch types, and it gives some pretty nice safety with few explicit types: https://github.com/sqren/kibana/blob/65a93973f823741e48894aa2875b819f03fb0a79/x-pack/plugins/apm/server/lib/traces/get_trace.ts#L62-L65

jasonrhodes replied:
I can do that, but I don't understand the difference?

jasonrhodes replied:
Oh you're saying to _replace_ the custom query result type with the ES type.

sqren replied:
I don't think `prepareTransactionGroups` should care about the structure of the ES response. All it cares about is the buckets. 

I think we should be careful not overtyping everything. When we make an interface typescript will blindly trust us, and I think it mostly makes sense to have interfaces when they are used multiple places. 

In this case we can get rid of `ITransactionGroupQueryResult` if it only takes the buckets.

sqren commented on x-pack/plugins/apm/server/lib/helpers/transaction_query.ts:
```diff
@@ -0,0 +1,119 @@
+      name: string;
+    };
+  };
+}
```
We should extract transaction and context from [here](https://github.com/elastic/kibana/blob/apm/dt/x-pack/plugins/apm/typings/Transaction.ts) and then do something like:

```
interface ITransactionGroupSample {
  transaction: TransactionTransaction;
  context: TransactionContext;
}
```

sqren replied:
Btw. I'm very open to better names than `TransactionTransaction ` :D

sqren replied:
... even `TransactionDocTransaction` is better but that would require us to rename the `Transaction` interface to `TransactionDoc`

sqren commented:
If you rebase with `apm/dt` you'll get `@types/elasticsearch` which might be useful.

jasonrhodes commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4035/)

jasonrhodes commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4046/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4043/)

jasonrhodes commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4069/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4094/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4113/)

sqren commented on x-pack/plugins/apm/common/constants.ts:
```diff
@@ -17,6 +17,10 @@ export const TRANSACTION_RESULT = 'transaction.result';
 export const TRANSACTION_NAME = 'transaction.name';
 export const TRANSACTION_ID = 'transaction.id';
 export const TRANSACTION_SAMPLED = 'transaction.sampled';
+export const TRANSACTION_SERVICE_NAME = 'context.service.name';
```
Is this resolved?

jasonrhodes replied:
Hm must have been reverted during rebase, re-removed it.

sqren commented on x-pack/plugins/apm/server/lib/traces/get_top_traces.ts:
```diff
@@ -0,0 +1,71 @@
+            exists: {
+              field: TRACE_ID
+            }
+          },
```
Errors can also have `trace_id`, so we should probably make sure it's a transaction:
`{ term: { [PROCESSOR_EVENT]: 'transaction' } }`

jasonrhodes replied:
Oh ok. Small confusion here: we don't currently do this same check in `get_top_transactions` (we seem to check for a single transaction type instead, perhaps?) and the get_error_groups query searches on a different index `config.get('apm_oss.errorIndices')`. What am I missing here and what's `PROCESSOR_EVENT`?

sqren replied:
> we don't currently do this same check in get_top_transactions (we seem to check for a single transaction type instead, perhaps?) 

Yes, error don't have `transaction.type` field. But to make it obvious we should probably add `processor.event` there as well.

> get_error_groups query searches on a different index config.get('apm_oss.errorIndices')

The default index is `apm-*`:
https://github.com/elastic/kibana/blob/master/src/core_plugins/apm_oss/index.js#L33

Which is why we add:`
`{ term: { [PROCESSOR_EVENT]: 'error' } }`

sqren replied:
> What am I missing here and what's PROCESSOR_EVENT

Processor event is basically the document type. Valid types are `transaction`, `span` and `error`

sqren commented on yarn.lock:
```diff
@@ -876,6 +876,10 @@ angular-mocks@1.4.7:
   version "1.4.7"
   resolved "https://registry.yarnpkg.com/angular-mocks/-/angular-mocks-1.4.7.tgz#d7343ee0a033f9216770bda573950f6814d95227"
 
+angular-paging@2.2.1:
```
What is this?

sqren commented on package.json:
```diff
@@ -69,6 +69,7 @@
     "@kbn/pm": "link:packages/kbn-pm",
     "@kbn/test-subj-selector": "link:packages/kbn-test-subj-selector",
     "@kbn/ui-framework": "link:packages/kbn-ui-framework",
+    "@types/elasticsearch": "^5.0.26",
```
We already have this in devDeps: https://github.com/elastic/kibana/blob/apm/dt/package.json#L225

jasonrhodes replied:
Forgot the `--dev` flag :headdesk:

sqren commented on x-pack/plugins/apm/server/lib/transactions/get_top_transactions.ts:
```diff
@@ -0,0 +1,63 @@
+  transactionType,
+  serviceName
+}: {
+  setup: StringMap<any>;
```
You could use the new `Setup` type. But don't worry about it now if you get a green build. We can always fix that later.

sqren commented on x-pack/plugins/apm/server/lib/helpers/transaction_group_query.ts:
```diff
@@ -0,0 +1,113 @@
+  const duration = moment.duration(end - start);
+  const minutes = duration.asMinutes();
+
+  const results = buckets.map((bucket: ITransactionGroupBucket) => {
```
Since `buckets` are typed, `bucket` should be inferred. But don't worry about that now.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4196/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4205/)

sqren commented on x-pack/plugins/apm/server/lib/transactions/get_top_transactions.ts:
```diff
@@ -0,0 +1,63 @@
+  const params = {
+    index: config.get('apm_oss.transactionIndices'),
+    body: {
+      size: 'banana',
```
@jasonrhodes This is causing:
```
{"statusCode":400,"error":"Bad Request","message":"[number_format_exception] For input string: \"banana\""}
```
